server/src/test/java/org/elasticsearch/cluster/coordination/ElectionSchedulerFactoryTests.java

```diff
@@ -0,0 +1,178 @@
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.cluster.coordination;
+
+import org.elasticsearch.common.lease.Releasable;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.common.unit.TimeValue;
+import org.elasticsearch.test.ESTestCase;
+import org.junit.Before;
+
+import static org.elasticsearch.cluster.coordination.ElectionSchedulerFactory.ELECTION_BACK_OFF_TIME_SETTING;
+import static org.elasticsearch.cluster.coordination.ElectionSchedulerFactory.ELECTION_MAX_TIMEOUT_SETTING;
+import static org.elasticsearch.cluster.coordination.ElectionSchedulerFactory.ELECTION_MIN_TIMEOUT_SETTING;
+import static org.elasticsearch.cluster.coordination.ElectionSchedulerFactory.validationExceptionMessage;
+import static org.elasticsearch.node.Node.NODE_NAME_SETTING;
+import static org.hamcrest.Matchers.greaterThanOrEqualTo;
+import static org.hamcrest.Matchers.is;
+import static org.hamcrest.Matchers.lessThanOrEqualTo;
+
+public class ElectionSchedulerFactoryTests extends ESTestCase {
+
+    private DeterministicTaskQueue deterministicTaskQueue;
+    private ElectionSchedulerFactory electionSchedulerFactory;
+    private boolean electionStarted = false;
+
+    @Before
+    public void createObjects() {
+        final Settings settings = Settings.builder().put(NODE_NAME_SETTING.getKey(), "node").build();
+        deterministicTaskQueue = new DeterministicTaskQueue(settings);
+        electionSchedulerFactory = new ElectionSchedulerFactory(settings, random(), deterministicTaskQueue.getThreadPool());
+    }
+
+    private TimeValue randomGracePeriod() {
+        return TimeValue.timeValueMillis(randomLongBetween(0, 10000));
+    }
+
+    private void startElection() {
+        assertFalse(electionStarted);
+        electionStarted = true;
+    }
+
+    private void assertElectionSchedule() {
+        final TimeValue initialGracePeriod = randomGracePeriod();
+
+        try (Releasable ignored = electionSchedulerFactory.startElectionScheduler(initialGracePeriod, this::startElection)) {
+            long lastElectionTime = deterministicTaskQueue.getCurrentTimeMillis();
+            int electionCount = 0;
+            while (true) {
+                electionCount++;
+
+                runElection();
+
+                final long thisElectionTime = deterministicTaskQueue.getCurrentTimeMillis();
+
+                if (electionCount == 1) {
+                    final long electionDelay = thisElectionTime - lastElectionTime;
+
+                    // Check grace period
+                    assertThat(electionDelay, greaterThanOrEqualTo(initialGracePeriod.millis()));
+
+                    // Check upper bound
+                    assertThat(electionDelay, lessThanOrEqualTo(ELECTION_MIN_TIMEOUT_SETTING.get(Settings.EMPTY).millis()
```

Comment:
should we inject random settings with min timeout/ max timeout/backoff to test this on different settings?

Reply:
Ok. This was a little painful because putting a `TimeValue` into a `Settings.Builder` converts it (lossily) to a string and then back again, and even rejects things with a fraction like `1.1m`. Putting a specific number of milliseconds in as a string is the answer.